### PR TITLE
Do not reject http req on websocket routes

### LINKED ISSUE #1752: Upgrade websocket request when header present, otherwise do not reject
The current websocket support appears to offer two choices:
- When route is configured with `use_websocket:true` upgrade if header is present, otherwise reject
- The spec doesn't say what the behavior is when `use_websocket:false`; are requests with the upgrade header rejected or is the header ignored and the request forwarded over http?

The behavior of Cloud Foundry's current edge proxy is:
- If the upgrade header is present, we initiate a TCP connection to the backend
- If the upgrade header is not present, we forward an HTTP request
- There is no user control of whether upgrade header is rejected or accepted
- Apps are responsible for rejecting connections/requests based on whether they require TCP or HTTP requests 

In order to transition from our current edge proxy to Envoy without downtime for upstream applications, we must continue supporting websockets on all routes without breaking apps that don't support them. 

Could we change the semantics of this boolean to:
- when `use_websocket:true` upgrade if header is present, otherwise forward request over http
- when `use_websocket:false` keep current behavior

### PR DESCRIPTION
**Pull request by rshriram**
Fixes #1752 
Also improved test coverage of Websocket code.

### CODE REVIEW

**Comment by rshriram** [comment]
fyi @ggreenway @shalako

**Comment by mattklein123** [comment]
@rshriram as long as you are in here can you fix the coverage for the WS code? There is some missing. Thank you.

**Comment by rshriram** [comment]
Yes, I was planning on doing that. Hence the [wip] thing..

**Comment by mattklein123** [comment]
@rshriram merge master once https://github.com/envoyproxy/envoy/pull/1794 goes in and you can get coverage results for your PR.

**Comment by shalako** [comment]
@grosenhouse

**Comment by shalako** [comment]
@rosenhouse

**Comment by rshriram** [comment]
ready for review @mattklein123 / @htuch / @alyssawilk / whoever is interested.

File: source/common/http/conn_manager_impl.h [hunk 1]
```diff
@@ -74,8 +74,7 @@ namespace Http {
   COUNTER(downstream_rq_tx_reset)                                                                  \
   COUNTER(downstream_rq_non_relative_path)                                                         \
   COUNTER(downstream_rq_ws_on_non_ws_route)                                                        \
-  COUNTER(downstream_rq_non_ws_on_ws_route)                                                        \
-  COUNTER(downstream_rq_too_large)                                                                 \
+  COUNTER(downstream_rq_too_large)                                                              \
```
Comment (alyssawilk) [inline source/common/http/conn_manager_impl.h]: Can you revert the change to downstream_rq_too_large so the \s line up?

File: test/common/http/conn_manager_impl_test.cc [hunk 2]
```diff
@@ -700,8 +751,7 @@ TEST_F(HttpConnectionManagerImplTest, WebSocketPrefixRewrite) {
   EXPECT_CALL(route_config_provider_.route_config_->route_->route_entry_,
               finalizeRequestHeaders(_, _));
   EXPECT_CALL(route_config_provider_.route_config_->route_->route_entry_, autoHostRewrite())
-      .WillOnce(Return(false));
-  // TODO (rshriram) figure out how to test the auto host rewrite. Need handle over headers.
```
Comment (alyssawilk) [inline test/common/http/conn_manager_impl_test.cc]: I don't see any behavior change with the switch from false to true.  How are we testing that this is performing as expected?

Comment (rshriram) [reply test/common/http/conn_manager_impl_test.cc]: Ah shoot. Good catch. Is there a way to get a handle over the request headers, so that I can check if the authority value has been properly rewritten?

Comment (alyssawilk) [reply test/common/http/conn_manager_impl_test.cc]: Are the headers passed into finalizeRequestHeaders the same object which is manipulated for autoHostRewrite?  You could try rewriting  finalizeRequestHeaders mock to Invoke a function which stashes a pointer and then check the headers both in the finalizeRequestHeaders for initial host and after autoHostRewrite for the new host.  I'm not sure it'd work - they might be moved/destroyed before you get a chance to inspect but it's worth a shot :-)

Comment (mattklein123) [reply test/common/http/conn_manager_impl_test.cc]: You create the headers and "decode" them below. This header map is stored inside the conn man and forwarded on as the request headers. Just store a raw pointer to it before you decode and then you can inspect the request headers whenever you like.

File: source/common/http/conn_manager_impl.h [hunk 3]
```diff
@@ -74,7 +74,6 @@ namespace Http {
   COUNTER(downstream_rq_tx_reset)                                                                  \
   COUNTER(downstream_rq_non_relative_path)                                                         \
   COUNTER(downstream_rq_ws_on_non_ws_route)                                                        \
-  COUNTER(downstream_rq_non_ws_on_ws_route)                                                        \
```
Comment (mattklein123) [inline source/common/http/conn_manager_impl.h]: Is there a doc update needed here? If not, can you add the WS stats to the conn manager docs?

**Comment by ggreenway** [comment]
👍

**Comment by rshriram** [comment]
100% codecov except for some dummy function in ws_handler_impl.h
https://2273-65214191-gh.circle-artifacts.com/0/build/envoy/generated/coverage/coverage.source_common_http_websocket_ws_handler_impl.cc.html

**Comment by shalako** [comment]
Thank you, @rshriram !
